File: src/micrsoft-agent-framewokr-agt-integration/dotnet/AgtMaf/Display.cs
namespace AgtMaf;

/// <summary>
/// Small ANSI-colour helpers for a clean, readable demo trace. Colour can be
/// disabled by setting the environment variable <c>AGT_MAF_NO_COLOR=1</c>.
/// </summary>
public static class Display
{
    private static readonly bool NoColor =
        Environment.GetEnvironmentVariable("AGT_MAF_NO_COLOR") is { } v &&
        v is not ("" or "0" or "false" or "False");

    public static string Reset => NoColor ? "" : "\u001b[0m";
    public static string Bold => NoColor ? "" : "\u001b[1m";
    public static string Dim => NoColor ? "" : "\u001b[2m";
    public static string Red => NoColor ? "" : "\u001b[31m";
    public static string Green => NoColor ? "" : "\u001b[32m";
    public static string Yellow => NoColor ? "" : "\u001b[33m";
    public static string Blue => NoColor ? "" : "\u001b[34m";
    public static string Cyan => NoColor ? "" : "\u001b[36m";
    public static string Grey => NoColor ? "" : "\u001b[90m";

    private const int Width = 78;

    public static void Banner(string title, string? subtitle = null)
    {
        Console.WriteLine($"\n{Bold}{Blue}{new string('=', Width)}{Reset}");
        Console.WriteLine($"{Bold}{Blue}  {title}{Reset}");
        if (subtitle is not null)
        {
            Console.WriteLine($"{Grey}  {subtitle}{Reset}");
        }
        Console.WriteLine($"{Bold}{Blue}{new string('=', Width)}{Reset}");
    }

    public static void Section(string title)
    {
        var pad = Math.Max(0, Width - title.Length - 4);
        Console.WriteLine($"\n{Bold}{Cyan}-- {title} {new string('-', pad)}{Reset}");
    }

    public static void User(string text) => Console.WriteLine($"{Bold}[user]{Reset} {text}");

    public static void Intercept(string layer, string target) =>
        Console.WriteLine($"{Bold}{Cyan}AGT intercept{Reset} {Dim}[{layer}]{Reset} -> {Bold}{target}{Reset}");

    public static void Allowed(string reason) => Console.WriteLine($"{Green}   ALLOW{Reset} {Grey}{reason}{Reset}");

    public static void Denied(string reason) => Console.WriteLine($"{Red}   DENY{Reset}  {reason}");

    public static void Escalate(string reason) => Console.WriteLine($"{Yellow}   ESCALATE{Reset} {reason}");

    public static void Info(string label, string value) => Console.WriteLine($"{Grey}   {label}:{Reset} {value}");

    public static void AgentSays(string text) => Console.WriteLine($"{Bold}{Green}[agent]{Reset} {text}");

    public static void Note(string text) => Console.WriteLine($"{Grey}   {text}{Reset}");
}

File: src/agent-governance-toolkit/microsoft-agent-framework/python/agt_maf/display.py
"""Small ANSI-colour helpers for a clean, readable demo trace.

These helpers render nicely both in a terminal and in Jupyter (which prints ANSI
escape codes in cell output). Colour can be disabled with ``AGT_MAF_NO_COLOR=1``.

Importing this module also best-effort switches stdout to UTF-8 so the emoji and
box-drawing characters in the demo trace render correctly on Windows consoles
(whose default code page is cp1252).
"""

from __future__ import annotations

import os
import sys

# Make the colourful trace safe on Windows terminals, whose default code page
# (cp1252) cannot encode the emoji / box-drawing glyphs. macOS and Linux default
# to UTF-8, so we only touch the streams on Windows. Guarded because some hosts
# (Jupyter, pytest capture) replace stdout with a stream that has no ``reconfigure``.
if sys.platform == "win32":
    for _stream in (sys.stdout, sys.stderr):
        try:
            _stream.reconfigure(encoding="utf-8")  # type: ignore[attr-defined]
        except (AttributeError, ValueError, OSError):
            pass

_NO_COLOR = os.getenv("AGT_MAF_NO_COLOR", "").strip() not in ("", "0", "false", "False")

# Core palette ------------------------------------------------------------
RESET = "" if _NO_COLOR else "\033[0m"
BOLD = "" if _NO_COLOR else "\033[1m"
DIM = "" if _NO_COLOR else "\033[2m"
RED = "" if _NO_COLOR else "\033[31m"
GREEN = "" if _NO_COLOR else "\033[32m"
YELLOW = "" if _NO_COLOR else "\033[33m"
BLUE = "" if _NO_COLOR else "\033[34m"
MAGENTA = "" if _NO_COLOR else "\033[35m"
CYAN = "" if _NO_COLOR else "\033[36m"
GREY = "" if _NO_COLOR else "\033[90m"

_WIDTH = 78


def banner(title: str, subtitle: str | None = None) -> None:
    """Print a top-level banner for a demo."""
    print(f"\n{BOLD}{BLUE}{'═' * _WIDTH}{RESET}")
    print(f"{BOLD}{BLUE}  {title}{RESET}")
    if subtitle:
        print(f"{GREY}  {subtitle}{RESET}")
    print(f"{BOLD}{BLUE}{'═' * _WIDTH}{RESET}")


def section(title: str) -> None:
    """Print a section divider."""
    pad = max(0, _WIDTH - len(title) - 4)
    print(f"\n{BOLD}{CYAN}── {title} {'─' * pad}{RESET}")


def user(text: str) -> None:
    print(f"{BOLD}👤 user:{RESET} {text}")


def intercept(layer: str, target: str) -> None:
    """Show that AGT intercepted an outbound action inside the MAF pipeline."""
    print(f"{MAGENTA}🛡  AGT intercept{RESET} {DIM}[{layer}]{RESET} → {BOLD}{target}{RESET}")


def allowed(reason: str) -> None:
    print(f"{GREEN}   ✅ ALLOW{RESET} {GREY}{reason}{RESET}")


def denied(reason: str) -> None:
    print(f"{RED}   ⛔ DENY{RESET}  {reason}")


def escalate(reason: str) -> None:
    print(f"{YELLOW}   ⚠️  ESCALATE{RESET} {reason}")


def info(label: str, value: str) -> None:
    print(f"{GREY}   {label}:{RESET} {value}")


def agent_says(text: str) -> None:
    print(f"{BOLD}{GREEN}🤖 agent:{RESET} {text}")


def note(text: str) -> None:
    print(f"{GREY}   {text}{RESET}")
